apis/apiextensions/v1alpha1/ccrd/crd.go

```diff
@@ -42,130 +41,127 @@ const (
-}
 
-// ForCompositeResource derives the CustomResourceDefinition for a composite
-// resource from the supplied CompositeResourceDefinition.
-func ForCompositeResource(d *v1alpha1.CompositeResourceDefinition) Option {
-	return func(crd *v1beta1.CustomResourceDefinition) error {
-		spec, err := NewSpec(d.Spec.CRDSpecTemplate)
-		if err != nil {
-			return errors.Wrap(err, errNewSpec)
-		}
+	crd := &extv1.CustomResourceDefinition{
+		Spec: extv1.CustomResourceDefinitionSpec{
+			Scope:    extv1.ClusterScoped,
+			Group:    xrd.Spec.CRDSpecTemplate.Group,
+			Names:    xrd.Spec.CRDSpecTemplate.Names,
+			Versions: []extv1.CustomResourceDefinitionVersion{vr},
+		},
+	}
 
-		crd.SetName(d.GetName())
-		crd.SetLabels(d.GetLabels())
-		crd.SetAnnotations(d.GetAnnotations())
-		crd.SetOwnerReferences([]metav1.OwnerReference{meta.AsController(
-			meta.TypedReferenceTo(d, v1alpha1.CompositeResourceDefinitionGroupVersionKind),
-		)})
-		crd.Spec.AdditionalPrinterColumns = CompositeResourcePrinterColumns()
-
-		crd.Spec.Group = spec.Group
-		crd.Spec.Version = spec.Version
-		crd.Spec.Names = spec.Names
-		crd.Spec.Names.Categories = append(crd.Spec.Names.Categories, CategoryComposite)
-		crd.Spec.AdditionalPrinterColumns = append(crd.Spec.AdditionalPrinterColumns, spec.AdditionalPrinterColumns...)
-		for k, v := range getSpecProps(spec) {
-			crd.Spec.Validation.OpenAPIV3Schema.Properties["spec"].Properties[k] = v
-		}
+	crd.SetName(xrd.GetName())
+	crd.SetLabels(xrd.GetLabels())
+	crd.SetAnnotations(xrd.GetAnnotations())
+	crd.SetOwnerReferences([]metav1.OwnerReference{meta.AsController(
+		meta.TypedReferenceTo(xrd, v1alpha1.CompositeResourceDefinitionGroupVersionKind),
+	)})
 
-		crd.Spec.Scope = v1beta1.ClusterScoped
-		for k, v := range CompositeResourceSpecProps() {
-			crd.Spec.Validation.OpenAPIV3Schema.Properties["spec"].Properties[k] = v
-		}
-		for k, v := range CompositeResourceStatusProps() {
-			crd.Spec.Validation.OpenAPIV3Schema.Properties["status"].Properties[k] = v
-		}
+	crd.Spec.Names.Categories = append(crd.Spec.Names.Categories, CategoryComposite)
 
-		return nil
-	}
+	return crd, nil
 }
 
 // ForCompositeResourceClaim derives the CustomResourceDefinition for a
 // composite resource claim from the supplied CompositeResourceDefinition.
-func ForCompositeResourceClaim(d *v1alpha1.CompositeResourceDefinition) Option {
-	return func(crd *v1beta1.CustomResourceDefinition) error {
-		spec, err := NewSpec(d.Spec.CRDSpecTemplate)
-		if err != nil {
-			return errors.Wrap(err, errNewSpec)
-		}
+func ForCompositeResourceClaim(xrd *v1alpha1.CompositeResourceDefinition) (*extv1.CustomResourceDefinition, error) {
+	if err := validateClaimNames(xrd); err != nil {
+		return nil, errors.Wrap(err, errInvalidClaimNames)
+	}
 
-		if err := validateClaimNames(d); err != nil {
-			return errors.Wrap(err, errInvalidClaimNames)
-		}
+	vr := extv1.CustomResourceDefinitionVersion{
+		Name:                     xrd.Spec.CRDSpecTemplate.Version,
```

Comment:
Are we keeping this? The `v1.CRD`s have only `versions` array instead of `version` as opposed to our `CompositeResourceDefinition`. I think now could be a good time for a breaking change here and adopt the current v1 CRD schema since we plan on making XRD `v1beta1`.

Reply:
Agreed, although I believe that is tracked in a separate issue so maybe a follow-up to this PR?

Reply:
I don't think we currently have an issue tracking this for XRs (just for Crossplane CRs in general). I'm actually not quite convinced that we should (or perhaps rather can cleanly) support serving multiple versions of an XR, so I'd like to raise a distinct issue to discuss the topic and get some community feedback.

Reply:
Issue raised: https://github.com/crossplane/crossplane/issues/1871